components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt

```diff
@@ -259,27 +270,22 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
                                     )
                                     val client = connector.connect()
                                     client.clientClosed.advise(connectionLifetime) {
-                                        thisLogger().d("clientClosed, ${client.prettyPrint()}")
                                         application.invokeLater {
-                                            connectionLifetime.onTermination {
-                                                thisLogger().d("clientTermination, clientClosed didChange fired, ${client.prettyPrint()}")
-                                                dispatcher.multicaster.didChange()
-                                            }
                                             connectionLifetime.terminate()
                                         }
                                     }
                                     client.onClientPresenceChanged.advise(connectionLifetime) {
-                                        thisLogger().d("presenceChanged, ${client.prettyPrint()}")
                                         application.invokeLater {
-                                            thisLogger().d("clientPresenceChanged didChange fired, ${client.prettyPrint()}")
-                                            dispatcher.multicaster.didChange()
                                             if (client.clientPresent) {
                                                 statusMessage.text = ""
                                             }
                                         }
                                     }
                                     thinClient = client
-                                    jetbrainsClientMap[update.workspaceId] = client
+                                    setClient(update.workspaceId, client)
+                                    connectionLifetime.onTermination {
```

Comment:
I wonder whether it can race that by now it is already terminated. Would not it be better to use `onTerminationOrNow`? Would will happen if `onTerminationOrNow` through not alive exception?

Reply:
> I wonder whether it can race that by now it is already terminated. Would not it be better to use `onTerminationOrNow`?

How about using `client.lifetime` instead of `connectionLifetime`? I think it might be more actuate to use each ThinClient's `lifetime` to handle termination event. I think there is no such race condition for `client.lifetime`
```suggestion
                                    client.lifetime.onTerminationOrNow { setClient(update.workspaceId, null) }
```

> Would will happen if `onTerminationOrNow` yield not alive exception?

I think `setClient(update.workspaceId, null)` in the nearest catch block would be fine. What do you think?